src/chainrules.jl

```diff
@@ -1,23 +1,33 @@
-import ChainRules.rrule
+import ChainRulesCore: rrule, no_rrule
 
 
-function chainrules_supported_signatures()
-    rrule_methods = methods(rrule).ms
+function function_signatures(fn)
+    rrule_methods = methods(fn).ms
     sigs = [remove_first_parameter(rr.sig) for rr in rrule_methods]
     # add keyword version of these functions as well
     kw_sigs = [kwsig for kwsig in map(kwfunc_signature, sigs) if kwsig !== Tuple{}]
     return [sigs; kw_sigs]
 end
 
 
-const CHAIN_RULE_PRIMITIVES = Ref(FunctionResolver{Bool}())
+const CHAINRULES_PRIMITIVES = Ref(FunctionResolver{Bool}())
+const NUM_CHAINRULES_METHODS = Ref{Int}(0)
 
 
-function update_chainrules_primitives!()
-    P = FunctionResolver{Bool}([sig => true for sig in chainrules_supported_signatures()])
-    delete!(P.signatures, Symbol(Any))
-    CHAIN_RULE_PRIMITIVES[] = P
+function update_chainrules_primitives!(;force=false)
+    num_methods = length(methods(rrule)) + length(methods(no_rrule))
+    if force || num_methods != NUM_CHAINRULES_METHODS[]
+        sigs_flags = [
+            [sig => true for sig in function_signatures(rrule)];
+            [sig => false for sig in function_signatures(no_rrule)]  # override rrule(sig...)
+            ]
+        @warn "Updating chainrules"
+        P = FunctionResolver{Bool}(sigs_flags)
+        delete!(P.signatures, Symbol(Any))
```

Comment:
Is this still needed?
The `rrule(Any...)` has a matching `no_rrule(Any...)`.
(Other than of optimization reasons, we would have implemeneted it as `@opt_out rrule(Any...)`)

Reply:
Yeah, I believe we can remove it.